### PR TITLE
Fix calculation for multiple skill levels

### LINKED ISSUE #13: Investigate if skill levels work correctly and fix
## Description
Investigate why the skill levels don't work correctly and fix it.
Message Loic when it works again so that he can inform Niklas.

## Acceptance Criteria
- [ ] skill levels work
- [ ] message Loic that it works again

### PR DESCRIPTION
**Pull request by myluki2000**
## Description of changes
Previously when there were multiple assessments in a chapter with different skill levels the user would gain points for, the calculation wouldn't be correct because the calculation assumed all skill levels were of the same type to calculate the maximum number of points the user could earn.

  
## How has this been tested?
Please describe the test strategy you followed.

- [x] automated unit test
- [ ] automated integration test
- [ ] automated acceptance test
- [ ] manual, exploratory test
    
## Checklist before requesting a review
- [ ] My code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My code fulfills all acceptance criteria
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] The test coverage (line and branch) is reasonably high, especially on Service classes and other classes with complex logic
- [ ] I have made corresponding changes to the documentation
- [ ] I have added explanation of architectural decision and rationales to [wiki/adr](https://github.com/IT-REX-Platform/wiki/tree/main/adr)
- [ ] I have updated the changes in the ticket description

## Checklist for reviewer
- [ ] The code works and does not throw errors
- [ ] The code is easy to understand and there are no confusing parts
- [ ] The code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [ ] The code change accomplishes what it is supposed to do
- [ ] I cannot think of any use case in which the code does not behave as intended
- [ ] The added and existing tests reasonably cover the code change
- [ ] I cannot think of any test cases, input or edge cases that should be tested in addition
- [ ] Description of the change is included in the documentation

### CODE REVIEW

File: src/main/java/de/unistuttgart/iste/gits/skilllevel_service/service/calculation/SkillLevelCalculator.java [hunk 1]
```diff
@@ -139,15 +128,24 @@ private void addLogItemsToSkillLevelsForAssessment(AllSkillLevelsEntity allSkill
             // add the log entry to the skill level. For now only add the skill points earned ("difference")
             // because the order of the log will not be correct for now because we go through each assessment
             // one by one. Later we will sort the log entries by the timestamp and calculate the missing values
-            for(SkillLevelEntity skillLevelToModify : skillLevelsToModify) {
+            for(SkillType skillType : skillTypes) {
+                // each chapter has a maximum of 10 skill levels, so we need to scale the earned skill points relative
+                // to the total skill points of the chapter times the 10 levels to calculate how many levels the user
+                // will gain
+                float relativeSkillPoints
```
Comment (PaulBredl) [inline src/main/java/de/unistuttgart/iste/gits/skilllevel_service/service/calculation/SkillLevelCalculator.java:135]: warum benutzt du keine doubles? is mir egal, aber halt eher ungewöhnlich :D

File: src/main/java/de/unistuttgart/iste/gits/skilllevel_service/service/calculation/SkillLevelCalculator.java [hunk 2]
```diff
@@ -287,4 +295,24 @@ public AssessmentRepetition(OffsetDateTime timestamp, float earnedSkillPoints) {
             this.earnedSkillPoints = earnedSkillPoints;
         }
     }
+
+    /**
+     * Helper class which can be used to track skill points of all types. Should not be used outside of this
+     * class because it has no encapsulation. Only used inside this parent class to make calculation methods more tidy.
+     */
+    private static class AllSkillPoints {
```
Comment (PaulBredl) [inline src/main/java/de/unistuttgart/iste/gits/skilllevel_service/service/calculation/SkillLevelCalculator.java:303]: dA hÄtTe MaN jEtZt ScHoN eIn ReCoRd NeHmEn KöNnEn

Comment (PaulBredl) [reply src/main/java/de/unistuttgart/iste/gits/skilllevel_service/service/calculation/SkillLevelCalculator.java:303]: Das spiegelt die Sonar QUBE Meinung wieder, nicht meine xD
Des überlass ich dir was du damit machst
image

**Review by PaulBredl — APPROVED** [review]
die test hab ich mir nicht angeschaut, wird schon passen was du da treibst

File: src/main/java/de/unistuttgart/iste/gits/skilllevel_service/service/calculation/SkillLevelCalculator.java (hunk 2 above)
Comment (myluki2000) [reply src/main/java/de/unistuttgart/iste/gits/skilllevel_service/service/calculation/SkillLevelCalculator.java:303]: > dA hÄtTe MaN jEtZt ScHoN eIn ReCoRd NeHmEn KöNnEn

Ne weil bei nem record wären die felder final.

Ich brauch halt die Möglichkeit die zu modifizieren. Ich weiß dass SQ rummeckert aber für ne *interne* (private) Klasse finde ich das in Ordnung. Die dient ja nur dafür, dass ich nicht meine Methoden mit 4mal so vielen Variablen vollspammen muss weil ich eine für jeden Skill Type bräuchte
